tests/xharness/Hardware/Simulators.cs

```diff
@@ -0,0 +1,416 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using Xamarin;
+using xharness.Collections;
+using Xharness;
+using Xharness.Execution;
+using Xharness.Logging;
+
+namespace xharness.Hardware {
+	public class Simulators : ISimulatorsLoader {
+		public IHarness Harness { get; set; }
+		public IProcessManager ProcessManager { get; set; } = new ProcessManager ();
+
+		bool loaded;
+		readonly SemaphoreSlim semaphore = new SemaphoreSlim (1);
+
+		BlockingEnumerableCollection<SimRuntime> supported_runtimes = new BlockingEnumerableCollection<SimRuntime> ();
+		BlockingEnumerableCollection<SimDeviceType> supported_device_types = new BlockingEnumerableCollection<SimDeviceType> ();
+		BlockingEnumerableCollection<SimDevice> available_devices = new BlockingEnumerableCollection<SimDevice> ();
+		BlockingEnumerableCollection<SimDevicePair> available_device_pairs = new BlockingEnumerableCollection<SimDevicePair> ();
+
+		public IEnumerable<SimRuntime> SupportedRuntimes => supported_runtimes;
+		public IEnumerable<SimDeviceType> SupportedDeviceTypes => supported_device_types;
+		public IEnumerable<SimDevice> AvailableDevices => available_devices;
+		public IEnumerable<SimDevicePair> AvailableDevicePairs => available_device_pairs;
+
+		public async Task LoadAsync (ILog log, bool include_locked = false, bool force = false)
+		{
+			await semaphore.WaitAsync ();
+			if (loaded) {
+				if (!force) {
+					semaphore.Release ();
+					return;
+				}
+				supported_runtimes.Reset ();
+				supported_device_types.Reset ();
+				available_devices.Reset ();
+				available_device_pairs.Reset ();
+			}
+			loaded = true;
+
+			await Task.Run (async () => {
+				var tmpfile = Path.GetTempFileName ();
+				try {
+					using (var process = new Process ()) {
+						process.StartInfo.FileName = Harness.MlaunchPath;
+						var arguments = new MlaunchArguments (
+							(MlaunchArgumentType.SdkRoot, Harness.XcodeRoot),
+							(MlaunchArgumentType.ListSim, tmpfile)
+						);
+						process.StartInfo.Arguments = arguments.AsCommandLine ();
+						log.WriteLine ("Launching {0} {1}", process.StartInfo.FileName, process.StartInfo.Arguments);
+						var rv = await ProcessManager.RunAsync (process, log, timeout: TimeSpan.FromSeconds (30));
```

Comment:
Not sure if part of this change but later this should also be taken care of by the `ProcessManager`

Reply:
yes, will come later. Had to cut the PR to not be a gazillion lines :/